aspnetcore/fundamentals/host/hosted-services/samples/3.x/BackgroundTasksSample/Services/BackgroundTaskQueue.cs

```diff
@@ -1,42 +1,38 @@
 using System;
-using System.Collections.Concurrent;
 using System.Threading;
+using System.Threading.Channels;
 using System.Threading.Tasks;
 
 namespace BackgroundTasksSample.Services
 {
     #region snippet1
     public interface IBackgroundTaskQueue
     {
-        void QueueBackgroundWorkItem(Func<CancellationToken, Task> workItem);
+        ValueTask QueueBackgroundWorkItemAsync(Func<CancellationToken, ValueTask> workItem);
 
-        Task<Func<CancellationToken, Task>> DequeueAsync(
+        ValueTask<Func<CancellationToken, ValueTask>> DequeueAsync(
             CancellationToken cancellationToken);
     }
 
     public class BackgroundTaskQueue : IBackgroundTaskQueue
     {
-        private ConcurrentQueue<Func<CancellationToken, Task>> _workItems = 
-            new ConcurrentQueue<Func<CancellationToken, Task>>();
-        private SemaphoreSlim _signal = new SemaphoreSlim(0);
+        private readonly Channel<Func<CancellationToken, ValueTask>> _queue = Channel.CreateUnbounded<Func<CancellationToken, ValueTask>>();
```

Comment:
We'll want to create a bounded channel here since people will copy and paste this into their apps. Add a comment about how it should be set based on the amount of concurrent jobs you want to allow in the queue concurrently.

Reply:
good point, thanks

Reply:
I slightly modified the Composition Root'scode, reading the capacity from the app settings and injecting it. Let me know if it's fine for you.